{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Prototype notebook to semantically represent Learning Path with Schema.org\n",
    "\n",
    "authors: Phil Reed, Alban Gaignard, Leyla Jael Castro\n",
    "\n",
    "- Initially drafted the 20th of November 2025, as part of Bioschemas activities and ELIXIR BioHackathon Europe 2025.\n",
    "- Refined at de.NBI BioHackathon Germany 2025.\n",
    "- To be revised by the community.\n",
    "\n",
    "Any questions, please contact phil.reed@manchester.ac.uk, alban.gaignard@univ-nantes.fr\n",
    "\n"
   ]
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:48.963125Z",
     "start_time": "2025-12-04T12:27:47.856239Z"
    }
   },
   "cell_type": "code",
   "source": "!pip install rdflib",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\r\n",
      "Requirement already satisfied: rdflib in /home/baum/.local/lib/python3.10/site-packages (7.5.0)\r\n",
      "Requirement already satisfied: isodate<1.0.0,>=0.7.2 in /home/baum/.local/lib/python3.10/site-packages (from rdflib) (0.7.2)\r\n",
      "Requirement already satisfied: pyparsing<4,>=2.1.0 in /usr/lib/python3/dist-packages (from rdflib) (2.4.7)\r\n",
      "\r\n",
      "\u001B[1m[\u001B[0m\u001B[34;49mnotice\u001B[0m\u001B[1;39;49m]\u001B[0m\u001B[39;49m A new release of pip is available: \u001B[0m\u001B[31;49m24.3.1\u001B[0m\u001B[39;49m -> \u001B[0m\u001B[32;49m25.3\u001B[0m\r\n",
      "\u001B[1m[\u001B[0m\u001B[34;49mnotice\u001B[0m\u001B[1;39;49m]\u001B[0m\u001B[39;49m To update, run: \u001B[0m\u001B[32;49mpython3 -m pip install --upgrade pip\u001B[0m\r\n"
     ]
    }
   ],
   "execution_count": 36
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Sample Learning path\n",
    "https://tess.elixir-europe.org/learning_paths/introduction-to-galaxy-and-sequence-analysis-6384c0ed-3546-41cf-ac30-bff8680dd96c\n",
    "\n",
    "LP structure: \n",
    "\n",
    "**Introduction to Galaxy and Sequence analysis** [syllabusSections=M1,M2]\n",
    "- **Module 1: Introduction to Galaxy** [itemListElement=11,12] [nextItem=M2]\n",
    "  - (1.1) A short introduction to Galaxy [nextItem=12]\n",
    "  - (1.2) Galaxy Basics for genomics [nextItem=21]\n",
    "- **Module 2: Basics of Genome Sequence Analysis** [itemListElement=21,22,23,24]\n",
    "  - (2.1) Quality Control [nextItem=22]\n",
    "  - (2.2) Mapping [nextItem=23]\n",
    "  - (2.3) An Introduction to Genome Assembly [nextItem=24]\n",
    "  - (2.4) Chloroplast genome assembly \n",
    "\n"
   ]
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.071101Z",
     "start_time": "2025-12-04T12:27:49.047166Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from rdflib import Graph\n",
    "\n",
    "my_learning_path = \"\"\"\n",
    "@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n",
    "@prefix owl: <http://www.w3.org/2002/07/owl#> .\n",
    "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n",
    "@prefix schema: <https://schema.org/> .\n",
    "@prefix dct: <http://purl.org/dc/terms/> .\n",
    "@prefix ex: <http://example.org/> .\n",
    "ex:GA_learning_path a schema:Course ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/LearningPath> ;\n",
    "    schema:name \"Introduction to Galaxy and Sequence analysis\" ;\n",
    "    schema:description \"This learning path aims to teach you the basics of Galaxy and analysis of sequencing data. \" ;\n",
    "    schema:provider ex:ExampleUniversity ;\n",
    "    schema:courseCode \"GSA101\" ;\n",
    "    schema:syllabusSections ex:Module_2, ex:Module_1 .\n",
    "\t\n",
    "ex:Module_1 a schema:Syllabus, schema:ListItem, schema:ItemList ;  \n",
    "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
    "    schema:name \"Module 1: Introduction to Galaxy\" ;\n",
    "    schema:teaches \"Learn how to create a workflow\" ;\n",
    "    schema:nextItem ex:Module_2 ;\n",
    "    schema:itemListElement ex:TM11, ex:TM12 .\n",
    "ex:TM11 a schema:LearningResource, schema:ListItem ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
    "    schema:name \"(1.1) A short introduction to Galaxy\" ;\n",
    "    schema:description \"What is Galaxy\" ;\n",
    "    schema:url \"https://tess.elixir-europe.org/materials/hands-on-for-a-short-introduction-to-galaxy-tutorial?lp=1%3A1\" ;\n",
    "    schema:nextItem ex:TM12 .\n",
    "ex:TM12 a schema:LearningResource, schema:ListItem ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
    "    schema:name \"(1.2) Galaxy Basics for genomics\" ;\n",
    "    schema:description \"The basic details for Galaxy genomics\" ;\n",
    "    schema:nextItem ex:TM21 .\n",
    "\n",
    "ex:Module_2 a schema:Syllabus, schema:ListItem, schema:ItemList  ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
    "    schema:name \"Module 2: Basics of Genome Sequence Analysis\" ;\n",
    "    schema:teaches \"Use genome browser to understand your data\" ;\n",
    "    schema:itemListElement ex:TM21, ex:TM22, ex:TM23, ex:TM24 .\n",
    "ex:TM21 a schema:LearningResource, schema:ListItem ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
    "    schema:name \"(2.1) Quality Control\" ;\n",
    "    schema:description \"Ensure you follow quality control practices\" ;\n",
    "    schema:nextItem ex:TM22 .\n",
    "ex:TM22 a schema:LearningResource, schema:ListItem ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
    "    schema:name \"(2.2) Mapping\" ;\n",
    "    schema:description \"Alignment and mapping\" ;\n",
    "    schema:nextItem ex:TM23 .\n",
    "ex:TM23 a schema:LearningResource, schema:ListItem ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
    "    schema:name \"(2.3) An Introduction to Genome Assembly\" ;\n",
    "    schema:description \"Description of TrainingMaterial 23\" ;\n",
    "    schema:nextItem ex:TM24 .\n",
    "ex:TM24 a schema:LearningResource, schema:ListItem ;\n",
    "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
    "    schema:name \"(2.4) Chloroplast genome assembly\" ;\n",
    "    schema:description \"Description of TrainingMaterial 24\"  .\n",
    "\"\"\"\n",
    "\n",
    "kg = Graph()\n",
    "kg.parse(data=my_learning_path, format=\"turtle\")\n",
    "kg.serialize(destination=\"lp1.ttl\", format=\"turtle\")\n",
    "kg.serialize(destination=\"lp1.jsonld\", format=\"json-ld\")"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Graph identifier=Nf9d0a87abb2c4899a437caac18148a42 (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 37
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.111817Z",
     "start_time": "2025-12-04T12:27:49.103934Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(kg.serialize(format=\"turtle\"))\n",
    "print(len(kg))"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@prefix dct: <http://purl.org/dc/terms/> .\n",
      "@prefix ex: <http://example.org/> .\n",
      "@prefix schema: <https://schema.org/> .\n",
      "\n",
      "ex:GA_learning_path a schema:Course ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/LearningPath> ;\n",
      "    schema:courseCode \"GSA101\" ;\n",
      "    schema:description \"This learning path aims to teach you the basics of Galaxy and analysis of sequencing data. \" ;\n",
      "    schema:name \"Introduction to Galaxy and Sequence analysis\" ;\n",
      "    schema:provider ex:ExampleUniversity ;\n",
      "    schema:syllabusSections ex:Module_1,\n",
      "        ex:Module_2 .\n",
      "\n",
      "ex:Module_1 a schema:ItemList,\n",
      "        schema:ListItem,\n",
      "        schema:Syllabus ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
      "    schema:itemListElement ex:TM11,\n",
      "        ex:TM12 ;\n",
      "    schema:name \"Module 1: Introduction to Galaxy\" ;\n",
      "    schema:nextItem ex:Module_2 ;\n",
      "    schema:teaches \"Learn how to create a workflow\" .\n",
      "\n",
      "ex:TM11 a schema:LearningResource,\n",
      "        schema:ListItem ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
      "    schema:description \"What is Galaxy\" ;\n",
      "    schema:name \"(1.1) A short introduction to Galaxy\" ;\n",
      "    schema:nextItem ex:TM12 ;\n",
      "    schema:url \"https://tess.elixir-europe.org/materials/hands-on-for-a-short-introduction-to-galaxy-tutorial?lp=1%3A1\" .\n",
      "\n",
      "ex:Module_2 a schema:ItemList,\n",
      "        schema:ListItem,\n",
      "        schema:Syllabus ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
      "    schema:itemListElement ex:TM21,\n",
      "        ex:TM22,\n",
      "        ex:TM23,\n",
      "        ex:TM24 ;\n",
      "    schema:name \"Module 2: Basics of Genome Sequence Analysis\" ;\n",
      "    schema:teaches \"Use genome browser to understand your data\" .\n",
      "\n",
      "ex:TM12 a schema:LearningResource,\n",
      "        schema:ListItem ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
      "    schema:description \"The basic details for Galaxy genomics\" ;\n",
      "    schema:name \"(1.2) Galaxy Basics for genomics\" ;\n",
      "    schema:nextItem ex:TM21 .\n",
      "\n",
      "ex:TM21 a schema:LearningResource,\n",
      "        schema:ListItem ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
      "    schema:description \"Ensure you follow quality control practices\" ;\n",
      "    schema:name \"(2.1) Quality Control\" ;\n",
      "    schema:nextItem ex:TM22 .\n",
      "\n",
      "ex:TM22 a schema:LearningResource,\n",
      "        schema:ListItem ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
      "    schema:description \"Alignment and mapping\" ;\n",
      "    schema:name \"(2.2) Mapping\" ;\n",
      "    schema:nextItem ex:TM23 .\n",
      "\n",
      "ex:TM23 a schema:LearningResource,\n",
      "        schema:ListItem ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
      "    schema:description \"Description of TrainingMaterial 23\" ;\n",
      "    schema:name \"(2.3) An Introduction to Genome Assembly\" ;\n",
      "    schema:nextItem ex:TM24 .\n",
      "\n",
      "ex:TM24 a schema:LearningResource,\n",
      "        schema:ListItem ;\n",
      "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
      "    schema:description \"Description of TrainingMaterial 24\" ;\n",
      "    schema:name \"(2.4) Chloroplast genome assembly\" .\n",
      "\n",
      "\n",
      "63\n"
     ]
    }
   ],
   "execution_count": 38
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.244506Z",
     "start_time": "2025-12-04T12:27:49.228758Z"
    }
   },
   "cell_type": "code",
   "source": "print(kg.serialize(format=\"json-ld\"))",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "  {\n",
      "    \"@id\": \"http://example.org/Module_2\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/Syllabus\",\n",
      "      \"https://schema.org/ListItem\",\n",
      "      \"https://schema.org/ItemList\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/LearningPathModule\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/itemListElement\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM21\"\n",
      "      },\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM22\"\n",
      "      },\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM23\"\n",
      "      },\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM24\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"Module 2: Basics of Genome Sequence Analysis\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/teaches\": [\n",
      "      {\n",
      "        \"@value\": \"Use genome browser to understand your data\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/GA_learning_path\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/Course\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/LearningPath\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/courseCode\": [\n",
      "      {\n",
      "        \"@value\": \"GSA101\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/description\": [\n",
      "      {\n",
      "        \"@value\": \"This learning path aims to teach you the basics of Galaxy and analysis of sequencing data. \"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"Introduction to Galaxy and Sequence analysis\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/provider\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/ExampleUniversity\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/syllabusSections\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/Module_2\"\n",
      "      },\n",
      "      {\n",
      "        \"@id\": \"http://example.org/Module_1\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/TM11\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/LearningResource\",\n",
      "      \"https://schema.org/ListItem\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/description\": [\n",
      "      {\n",
      "        \"@value\": \"What is Galaxy\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"(1.1) A short introduction to Galaxy\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/nextItem\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM12\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/url\": [\n",
      "      {\n",
      "        \"@value\": \"https://tess.elixir-europe.org/materials/hands-on-for-a-short-introduction-to-galaxy-tutorial?lp=1%3A1\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/Module_1\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/Syllabus\",\n",
      "      \"https://schema.org/ListItem\",\n",
      "      \"https://schema.org/ItemList\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/LearningPathModule\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/itemListElement\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM11\"\n",
      "      },\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM12\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"Module 1: Introduction to Galaxy\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/nextItem\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/Module_2\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/teaches\": [\n",
      "      {\n",
      "        \"@value\": \"Learn how to create a workflow\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/TM23\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/LearningResource\",\n",
      "      \"https://schema.org/ListItem\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/description\": [\n",
      "      {\n",
      "        \"@value\": \"Description of TrainingMaterial 23\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"(2.3) An Introduction to Genome Assembly\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/nextItem\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM24\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/TM24\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/LearningResource\",\n",
      "      \"https://schema.org/ListItem\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/description\": [\n",
      "      {\n",
      "        \"@value\": \"Description of TrainingMaterial 24\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"(2.4) Chloroplast genome assembly\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/TM22\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/LearningResource\",\n",
      "      \"https://schema.org/ListItem\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/description\": [\n",
      "      {\n",
      "        \"@value\": \"Alignment and mapping\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"(2.2) Mapping\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/nextItem\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM23\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/TM12\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/LearningResource\",\n",
      "      \"https://schema.org/ListItem\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/description\": [\n",
      "      {\n",
      "        \"@value\": \"The basic details for Galaxy genomics\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"(1.2) Galaxy Basics for genomics\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/nextItem\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM21\"\n",
      "      }\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"@id\": \"http://example.org/TM21\",\n",
      "    \"@type\": [\n",
      "      \"https://schema.org/LearningResource\",\n",
      "      \"https://schema.org/ListItem\"\n",
      "    ],\n",
      "    \"http://purl.org/dc/terms/conformsTo\": [\n",
      "      {\n",
      "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/description\": [\n",
      "      {\n",
      "        \"@value\": \"Ensure you follow quality control practices\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/name\": [\n",
      "      {\n",
      "        \"@value\": \"(2.1) Quality Control\"\n",
      "      }\n",
      "    ],\n",
      "    \"https://schema.org/nextItem\": [\n",
      "      {\n",
      "        \"@id\": \"http://example.org/TM22\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "]\n"
     ]
    }
   ],
   "execution_count": 39
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.274433Z",
     "start_time": "2025-12-04T12:27:49.260003Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# I'm interested in a specific training material, what should be the prerequisites ?\n",
    "\n",
    "tm_of_interest = \"(2.3) An Introduction to Genome Assembly\"\n",
    "\n",
    "def query_learning_path():\n",
    "  query = f\"\"\"\n",
    "  SELECT * WHERE {{\n",
    "    ?s schema:name \"{tm_of_interest}\" .\n",
    "    ?prereq schema:nextItem+ ?s .\n",
    "    ?prereq schema:name ?prereq_name .\n",
    "    ?a_module schema:itemListElement ?prereq .\n",
    "    ?a_module schema:name ?module_name .\n",
    "  }}\n",
    "  \"\"\"\n",
    "  # Run SPARQL query\n",
    "  res_kg = kg.query(query)\n",
    "  print(f'If you are interested in {tm_of_interest}')\n",
    "  for r in res_kg:\n",
    "    print(f'You need to perform {r[\"prereq_name\"]} as part of the module {r[\"module_name\"]}')\n",
    "  return res_kg\n",
    "\n",
    "query_learning_path()"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "If you are interested in (2.3) An Introduction to Genome Assembly\n",
      "You need to perform (2.2) Mapping as part of the module Module 2: Basics of Genome Sequence Analysis\n",
      "You need to perform (2.1) Quality Control as part of the module Module 2: Basics of Genome Sequence Analysis\n",
      "You need to perform (1.2) Galaxy Basics for genomics as part of the module Module 1: Introduction to Galaxy\n",
      "You need to perform (1.1) A short introduction to Galaxy as part of the module Module 1: Introduction to Galaxy\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<rdflib.plugins.sparql.processor.SPARQLResult at 0x7ff69236b7c0>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 40
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## What we did until now\n",
    "\n",
    "1.  **RDF Graph Initialization**: We loaded a learning path described in Turtle format into an RDFLib graph (`kg`). We extensively used Schema.org to model the LearningPath Syllabus, ItemList and ListItem types, as well as Course and LearningResource types.\n",
    "2.  **Prerequisite Query**: We demonstrated how to query the graph using SPARQL to find prerequisites for a specific learning material.\n"
   ]
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Visualization of the Schema.org metadata"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.363838Z",
     "start_time": "2025-12-04T12:27:49.328971Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from rdflib import Graph, Literal, URIRef\n",
    "from rdflib.namespace import RDF, RDFS, OWL, XSD\n",
    "\n",
    "def extract_relationships_and_names(graph):\n",
    "    # Define prefixes\n",
    "    schema = URIRef(\"https://schema.org/\")\n",
    "    ex = URIRef(\"http://example.org/\")\n",
    "\n",
    "    # 1. Query for all entities with a schema:name and store their names\n",
    "    entity_names = {}\n",
    "    name_query = \"\"\"\n",
    "    PREFIX schema: <https://schema.org/>\n",
    "    SELECT ?entity ?name\n",
    "    WHERE {\n",
    "        ?entity schema:name ?name .\n",
    "    }\n",
    "    \"\"\"\n",
    "    for row in graph.query(name_query):\n",
    "        entity_names[str(row.entity)] = str(row.name)\n",
    "\n",
    "    print(\"Extracted entity names:\")\n",
    "    for uri, name in entity_names.items():\n",
    "        print(f\"  {uri}: {name}\")\n",
    "\n",
    "    # 2. Query for sequential relationships\n",
    "    relationships_query = \"\"\"\n",
    "    PREFIX schema: <https://schema.org/>\n",
    "    SELECT ?s ?p ?o\n",
    "    WHERE {\n",
    "        ?s ?p ?o .\n",
    "        FILTER (?p = schema:nextItem || ?p = schema:itemListElement || ?p = schema:step)\n",
    "    }\n",
    "    \"\"\"\n",
    "    raw_relationships = []\n",
    "    for row in graph.query(relationships_query):\n",
    "        raw_relationships.append((str(row.s), str(row.p), str(row.o)))\n",
    "\n",
    "    print(\"\\nExtracted raw relationships:\")\n",
    "    for rel in raw_relationships:\n",
    "        print(f\"  {rel[0]} -- {rel[1]} --> {rel[2]}\")\n",
    "\n",
    "    # 3. Transform relationships to use names\n",
    "    mermaid_edges = []\n",
    "    for s_uri, p_uri, o_uri in raw_relationships:\n",
    "        s_name = entity_names.get(s_uri, s_uri.split('/')[-1]) # Use URI fragment if name not found\n",
    "        o_name = entity_names.get(o_uri, o_uri.split('/')[-1]) # Use URI fragment if name not found\n",
    "        predicate_type = p_uri.split('/')[-1] # Get predicate name (e.g., nextItem)\n",
    "        mermaid_edges.append((s_name, predicate_type, o_name))\n",
    "\n",
    "    print(\"\\nTransformed relationships for Mermaid diagram:\")\n",
    "    for edge in mermaid_edges:\n",
    "        print(f\"  {edge[0]} -- {edge[1]} --> {edge[2]}\")\n",
    "\n",
    "    return entity_names, mermaid_edges\n",
    "\n",
    "# Call the function with the existing kg graph\n",
    "all_entity_names, all_mermaid_edges = extract_relationships_and_names(kg)\n",
    "\n",
    "# Assign to the global query_learning_path function as well if needed later\n",
    "def query_learning_path():\n",
    "    return all_entity_names, all_mermaid_edges\n",
    "\n",
    "print(\"Relationship extraction complete. Results stored in all_entity_names and all_mermaid_edges.\")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted entity names:\n",
      "  http://example.org/GA_learning_path: Introduction to Galaxy and Sequence analysis\n",
      "  http://example.org/Module_1: Module 1: Introduction to Galaxy\n",
      "  http://example.org/TM11: (1.1) A short introduction to Galaxy\n",
      "  http://example.org/TM12: (1.2) Galaxy Basics for genomics\n",
      "  http://example.org/Module_2: Module 2: Basics of Genome Sequence Analysis\n",
      "  http://example.org/TM21: (2.1) Quality Control\n",
      "  http://example.org/TM22: (2.2) Mapping\n",
      "  http://example.org/TM23: (2.3) An Introduction to Genome Assembly\n",
      "  http://example.org/TM24: (2.4) Chloroplast genome assembly\n",
      "\n",
      "Extracted raw relationships:\n",
      "  http://example.org/Module_1 -- https://schema.org/itemListElement --> http://example.org/TM12\n",
      "  http://example.org/Module_1 -- https://schema.org/itemListElement --> http://example.org/TM11\n",
      "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM24\n",
      "  http://example.org/TM11 -- https://schema.org/nextItem --> http://example.org/TM12\n",
      "  http://example.org/Module_1 -- https://schema.org/nextItem --> http://example.org/Module_2\n",
      "  http://example.org/TM23 -- https://schema.org/nextItem --> http://example.org/TM24\n",
      "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM21\n",
      "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM22\n",
      "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM23\n",
      "  http://example.org/TM22 -- https://schema.org/nextItem --> http://example.org/TM23\n",
      "  http://example.org/TM12 -- https://schema.org/nextItem --> http://example.org/TM21\n",
      "  http://example.org/TM21 -- https://schema.org/nextItem --> http://example.org/TM22\n",
      "\n",
      "Transformed relationships for Mermaid diagram:\n",
      "  Module 1: Introduction to Galaxy -- itemListElement --> (1.2) Galaxy Basics for genomics\n",
      "  Module 1: Introduction to Galaxy -- itemListElement --> (1.1) A short introduction to Galaxy\n",
      "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.4) Chloroplast genome assembly\n",
      "  (1.1) A short introduction to Galaxy -- nextItem --> (1.2) Galaxy Basics for genomics\n",
      "  Module 1: Introduction to Galaxy -- nextItem --> Module 2: Basics of Genome Sequence Analysis\n",
      "  (2.3) An Introduction to Genome Assembly -- nextItem --> (2.4) Chloroplast genome assembly\n",
      "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.1) Quality Control\n",
      "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.2) Mapping\n",
      "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.3) An Introduction to Genome Assembly\n",
      "  (2.2) Mapping -- nextItem --> (2.3) An Introduction to Genome Assembly\n",
      "  (1.2) Galaxy Basics for genomics -- nextItem --> (2.1) Quality Control\n",
      "  (2.1) Quality Control -- nextItem --> (2.2) Mapping\n",
      "Relationship extraction complete. Results stored in all_entity_names and all_mermaid_edges.\n"
     ]
    }
   ],
   "execution_count": 41
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.480284Z",
     "start_time": "2025-12-04T12:27:49.456984Z"
    }
   },
   "cell_type": "code",
   "source": [
    "mermaid_lines = []\n",
    "mermaid_lines.append('graph TD')\n",
    "\n",
    "for s_name, p_type, o_name in all_mermaid_edges:\n",
    "    # Sanitize names for Mermaid if they contain special characters or spaces\n",
    "    # Mermaid node IDs cannot contain spaces or special characters directly\n",
    "    # We'll use a simple approach: if a name has spaces, enclose it in quotes.\n",
    "    # For simplicity, we'll map original names to safe IDs and use names for labels\n",
    "\n",
    "    # Create a unique ID for each node if it's not already in the entity_names (which are already unique names)\n",
    "    # For this exercise, we will assume the names are unique enough for node IDs or simply use them as both ID and label.\n",
    "    # Mermaid allows IDs with spaces if quoted, but it's cleaner to have simple IDs.\n",
    "    # However, given the prompt directly asks for f'{source_name} -- {predicate_type} --> {target_name}',\n",
    "    # we will use the names as they are, and Mermaid should handle simple spaces correctly in labels.\n",
    "\n",
    "    # Let's ensure node labels are quoted if they contain spaces to be robust\n",
    "    # Mermaid syntax: A[\"Node Name\"] or A --> B[\"Another Node\"]\n",
    "    # For simplicity, let's use the names directly as node IDs and labels.\n",
    "    # Mermaid handles spaces in node labels gracefully if they are quoted, but for connections\n",
    "    # the IDs should be simple. Let's make sure our approach is compatible.\n",
    "\n",
    "    # A common Mermaid practice is to have a simple ID for the node and then a label\n",
    "    # like: ID[Label]. If we use the name as ID, it should not have spaces.\n",
    "    # Given the previous step produced names like 'Module 1', we should create a mapping for node IDs.\n",
    "\n",
    "    # Let's refine how we generate the Mermaid lines to ensure valid syntax\n",
    "    # Create a mapping from clean name to a unique, mermaid-safe ID\n",
    "    # This mapping is important to handle cases where names might have characters that Mermaid IDs don't like\n",
    "    node_id_map = {}\n",
    "    counter = 0\n",
    "    for uri_name in all_entity_names.values():\n",
    "        clean_name = uri_name.replace(' ', '_').replace('-', '_').replace(':', '_').replace('.', '_').strip()\n",
    "        if clean_name not in node_id_map.values():\n",
    "            node_id_map[uri_name] = f'N{counter}'\n",
    "            counter += 1\n",
    "\n",
    "    # Create mermaid lines with unique IDs and original names as labels\n",
    "    mermaid_lines = ['graph TD']\n",
    "    for s_name, p_type, o_name in all_mermaid_edges:\n",
    "        source_id = node_id_map.get(s_name, s_name.replace(' ', '_').replace('-', '_').replace(':', '_').replace('.', '_').strip())\n",
    "        target_id = node_id_map.get(o_name, o_name.replace(' ', '_').replace('-', '_').replace(':', '_').replace('.', '_').strip())\n",
    "\n",
    "        # Add node definitions with labels if they contain spaces\n",
    "        if ' ' in s_name:\n",
    "            mermaid_lines.append(f'{source_id}[\\\"{s_name}\\\"]')\n",
    "        if ' ' in o_name:\n",
    "            mermaid_lines.append(f'{target_id}[\\\"{o_name}\\\"]')\n",
    "\n",
    "        mermaid_lines.append(f'{source_id} -- {p_type} --> {target_id}')\n",
    "\n",
    "mermaid_syntax = '\\n'.join(mermaid_lines)\n",
    "\n",
    "# save mermaid file to disk\n",
    "with open(\"lp1.mmd\", \"w\") as mermaid_file:\n",
    "    print(mermaid_syntax, file=mermaid_file)\n",
    "\n",
    "# display the markdown mermaid graph as output of this cell\n",
    "print(mermaid_syntax)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "graph TD\n",
      "N1[\"Module 1: Introduction to Galaxy\"]\n",
      "N3[\"(1.2) Galaxy Basics for genomics\"]\n",
      "N1 -- itemListElement --> N3\n",
      "N1[\"Module 1: Introduction to Galaxy\"]\n",
      "N2[\"(1.1) A short introduction to Galaxy\"]\n",
      "N1 -- itemListElement --> N2\n",
      "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
      "N8[\"(2.4) Chloroplast genome assembly\"]\n",
      "N4 -- itemListElement --> N8\n",
      "N2[\"(1.1) A short introduction to Galaxy\"]\n",
      "N3[\"(1.2) Galaxy Basics for genomics\"]\n",
      "N2 -- nextItem --> N3\n",
      "N1[\"Module 1: Introduction to Galaxy\"]\n",
      "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
      "N1 -- nextItem --> N4\n",
      "N7[\"(2.3) An Introduction to Genome Assembly\"]\n",
      "N8[\"(2.4) Chloroplast genome assembly\"]\n",
      "N7 -- nextItem --> N8\n",
      "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
      "N5[\"(2.1) Quality Control\"]\n",
      "N4 -- itemListElement --> N5\n",
      "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
      "N6[\"(2.2) Mapping\"]\n",
      "N4 -- itemListElement --> N6\n",
      "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
      "N7[\"(2.3) An Introduction to Genome Assembly\"]\n",
      "N4 -- itemListElement --> N7\n",
      "N6[\"(2.2) Mapping\"]\n",
      "N7[\"(2.3) An Introduction to Genome Assembly\"]\n",
      "N6 -- nextItem --> N7\n",
      "N3[\"(1.2) Galaxy Basics for genomics\"]\n",
      "N5[\"(2.1) Quality Control\"]\n",
      "N3 -- nextItem --> N5\n",
      "N5[\"(2.1) Quality Control\"]\n",
      "N6[\"(2.2) Mapping\"]\n",
      "N5 -- nextItem --> N6\n"
     ]
    }
   ],
   "execution_count": 42
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.661487Z",
     "start_time": "2025-12-04T12:27:49.511026Z"
    }
   },
   "cell_type": "code",
   "source": [
    "#from IPython.display import display, Markdown\n",
    "#Markdown(mermaid_syntax)\n",
    "\n",
    "import base64\n",
    "from IPython.display import display_svg\n",
    "from urllib.request import Request, urlopen\n",
    "\n",
    "def mm(graph):\n",
    "    graphbytes = graph.encode(\"ascii\")\n",
    "    base64_bytes = base64.b64encode(graphbytes)\n",
    "    base64_string = base64_bytes.decode(\"ascii\")\n",
    "    url=\"https://mermaid.ink/svg/\" + base64_string\n",
    "    req=Request(url, headers={'User-Agent': 'IPython/Notebook'})\n",
    "    display_svg(urlopen(req).read().decode(), raw=True)\n",
    "mm(mermaid_syntax)"
   ],
   "outputs": [
    {
     "data": {
      "image/svg+xml": "<svg id=\"mermaidInkSvg\" width=\"100%\" xmlns=\"http://www.w3.org/2000/svg\" class=\"flowchart\" style=\"max-width: 845.765625px;\" viewBox=\"0 0 845.765625 982\" role=\"graphics-document document\" aria-roledescription=\"flowchart-v2\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><style xmlns=\"http://www.w3.org/1999/xhtml\">@import url(\"https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.7.2/css/all.min.css\");</style><style>#mermaidInkSvg{font-family:\"trebuchet ms\",verdana,arial,sans-serif;font-size:16px;fill:#333;}@keyframes edge-animation-frame{from{stroke-dashoffset:0;}}@keyframes dash{to{stroke-dashoffset:0;}}#mermaidInkSvg .edge-animation-slow{stroke-dasharray:9,5!important;stroke-dashoffset:900;animation:dash 50s linear infinite;stroke-linecap:round;}#mermaidInkSvg .edge-animation-fast{stroke-dasharray:9,5!important;stroke-dashoffset:900;animation:dash 20s linear infinite;stroke-linecap:round;}#mermaidInkSvg .error-icon{fill:#552222;}#mermaidInkSvg .error-text{fill:#552222;stroke:#552222;}#mermaidInkSvg .edge-thickness-normal{stroke-width:1px;}#mermaidInkSvg .edge-thickness-thick{stroke-width:3.5px;}#mermaidInkSvg .edge-pattern-solid{stroke-dasharray:0;}#mermaidInkSvg .edge-thickness-invisible{stroke-width:0;fill:none;}#mermaidInkSvg .edge-pattern-dashed{stroke-dasharray:3;}#mermaidInkSvg .edge-pattern-dotted{stroke-dasharray:2;}#mermaidInkSvg .marker{fill:#333333;stroke:#333333;}#mermaidInkSvg .marker.cross{stroke:#333333;}#mermaidInkSvg svg{font-family:\"trebuchet ms\",verdana,arial,sans-serif;font-size:16px;}#mermaidInkSvg p{margin:0;}#mermaidInkSvg .label{font-family:\"trebuchet ms\",verdana,arial,sans-serif;color:#333;}#mermaidInkSvg .cluster-label text{fill:#333;}#mermaidInkSvg .cluster-label span{color:#333;}#mermaidInkSvg .cluster-label span p{background-color:transparent;}#mermaidInkSvg .label text,#mermaidInkSvg span{fill:#333;color:#333;}#mermaidInkSvg .node rect,#mermaidInkSvg .node circle,#mermaidInkSvg .node ellipse,#mermaidInkSvg .node polygon,#mermaidInkSvg .node path{fill:#ECECFF;stroke:#9370DB;stroke-width:1px;}#mermaidInkSvg .rough-node .label text,#mermaidInkSvg .node .label text,#mermaidInkSvg .image-shape .label,#mermaidInkSvg .icon-shape .label{text-anchor:middle;}#mermaidInkSvg .node .katex path{fill:#000;stroke:#000;stroke-width:1px;}#mermaidInkSvg .rough-node .label,#mermaidInkSvg .node .label,#mermaidInkSvg .image-shape .label,#mermaidInkSvg .icon-shape .label{text-align:center;}#mermaidInkSvg .node.clickable{cursor:pointer;}#mermaidInkSvg .root .anchor path{fill:#333333!important;stroke-width:0;stroke:#333333;}#mermaidInkSvg .arrowheadPath{fill:#333333;}#mermaidInkSvg .edgePath .path{stroke:#333333;stroke-width:2.0px;}#mermaidInkSvg .flowchart-link{stroke:#333333;fill:none;}#mermaidInkSvg .edgeLabel{background-color:rgba(232,232,232, 0.8);text-align:center;}#mermaidInkSvg .edgeLabel p{background-color:rgba(232,232,232, 0.8);}#mermaidInkSvg .edgeLabel rect{opacity:0.5;background-color:rgba(232,232,232, 0.8);fill:rgba(232,232,232, 0.8);}#mermaidInkSvg .labelBkg{background-color:rgba(232, 232, 232, 0.5);}#mermaidInkSvg .cluster rect{fill:#ffffde;stroke:#aaaa33;stroke-width:1px;}#mermaidInkSvg .cluster text{fill:#333;}#mermaidInkSvg .cluster span{color:#333;}#mermaidInkSvg div.mermaidTooltip{position:absolute;text-align:center;max-width:200px;padding:2px;font-family:\"trebuchet ms\",verdana,arial,sans-serif;font-size:12px;background:hsl(80, 100%, 96.2745098039%);border:1px solid #aaaa33;border-radius:2px;pointer-events:none;z-index:100;}#mermaidInkSvg .flowchartTitleText{text-anchor:middle;font-size:18px;fill:#333;}#mermaidInkSvg rect.text{fill:none;stroke-width:0;}#mermaidInkSvg .icon-shape,#mermaidInkSvg .image-shape{background-color:rgba(232,232,232, 0.8);text-align:center;}#mermaidInkSvg .icon-shape p,#mermaidInkSvg .image-shape p{background-color:rgba(232,232,232, 0.8);padding:2px;}#mermaidInkSvg .icon-shape rect,#mermaidInkSvg .image-shape rect{opacity:0.5;background-color:rgba(232,232,232, 0.8);fill:rgba(232,232,232, 0.8);}#mermaidInkSvg .label-icon{display:inline-block;height:1em;overflow:visible;vertical-align:-0.125em;}#mermaidInkSvg .node .label-icon path{fill:currentColor;stroke:revert;stroke-width:revert;}#mermaidInkSvg :root{--mermaid-font-family:\"trebuchet ms\",verdana,arial,sans-serif;}</style><g><marker id=\"mermaidInkSvg_flowchart-v2-pointEnd\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"5\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"8\" markerHeight=\"8\" orient=\"auto\"><path d=\"M 0 0 L 10 5 L 0 10 z\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-pointStart\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"4.5\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"8\" markerHeight=\"8\" orient=\"auto\"><path d=\"M 0 5 L 10 10 L 10 0 z\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-circleEnd\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"11\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><circle cx=\"5\" cy=\"5\" r=\"5\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-circleStart\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"-1\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><circle cx=\"5\" cy=\"5\" r=\"5\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-crossEnd\" class=\"marker cross flowchart-v2\" viewBox=\"0 0 11 11\" refX=\"12\" refY=\"5.2\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><path d=\"M 1,1 l 9,9 M 10,1 l -9,9\" class=\"arrowMarkerPath\" style=\"stroke-width: 2; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-crossStart\" class=\"marker cross flowchart-v2\" viewBox=\"0 0 11 11\" refX=\"-1\" refY=\"5.2\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><path d=\"M 1,1 l 9,9 M 10,1 l -9,9\" class=\"arrowMarkerPath\" style=\"stroke-width: 2; stroke-dasharray: 1, 0;\"/></marker><g class=\"root\"><g class=\"clusters\"/><g class=\"edgePaths\"><path d=\"M462.211,86L466.151,92.167C470.091,98.333,477.971,110.667,481.912,129.5C485.852,148.333,485.852,173.667,485.852,199C485.852,224.333,485.852,249.667,495.626,270.086C505.401,290.505,524.951,306.01,534.726,313.762L544.5,321.514\" id=\"L_N1_N3_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N1_N3_0\" data-points=\"W3sieCI6NDYyLjIxMTE5NDQ5MDEzMTU2LCJ5Ijo4Nn0seyJ4Ijo0ODUuODUxNTYyNSwieSI6MTIzfSx7IngiOjQ4NS44NTE1NjI1LCJ5IjoxOTl9LHsieCI6NDg1Ljg1MTU2MjUsInkiOjI3NX0seyJ4Ijo1NDcuNjM0NDU0OTAwNTY4MSwieSI6MzI0fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M567.293,83.529L590.705,90.107C614.117,96.686,660.941,109.843,684.354,121.921C707.766,134,707.766,145,707.766,150.5L707.766,156\" id=\"L_N1_N2_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N1_N2_0\" data-points=\"W3sieCI6NTY3LjI5Mjk2ODc1LCJ5Ijo4My41Mjg2NDYzMjIyNjU3fSx7IngiOjcwNy43NjU2MjUsInkiOjEyM30seyJ4Ijo3MDcuNzY1NjI1LCJ5IjoxNjB9XQ==\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M129.592,414L118.812,420.167C108.033,426.333,86.473,438.667,75.694,455.5C64.914,472.333,64.914,493.667,64.914,515C64.914,536.333,64.914,557.667,64.914,579C64.914,600.333,64.914,621.667,64.914,643C64.914,664.333,64.914,685.667,64.914,709C64.914,732.333,64.914,757.667,64.914,783C64.914,808.333,64.914,833.667,83.887,852.3C102.86,870.933,140.806,882.867,159.779,888.833L178.752,894.8\" id=\"L_N4_N8_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N8_0\" data-points=\"W3sieCI6MTI5LjU5MTc5Njg3NSwieSI6NDE0fSx7IngiOjY0LjkxNDA2MjUsInkiOjQ1MX0seyJ4Ijo2NC45MTQwNjI1LCJ5Ijo1MTV9LHsieCI6NjQuOTE0MDYyNSwieSI6NTc5fSx7IngiOjY0LjkxNDA2MjUsInkiOjY0M30seyJ4Ijo2NC45MTQwNjI1LCJ5Ijo3MDd9LHsieCI6NjQuOTE0MDYyNSwieSI6NzgzfSx7IngiOjY0LjkxNDA2MjUsInkiOjg1OX0seyJ4IjoxODIuNTY4MjA1MTgwOTIxMDQsInkiOjg5Nn1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M707.766,238L707.766,244.167C707.766,250.333,707.766,262.667,697.991,276.586C688.216,290.505,668.666,306.01,658.892,313.762L649.117,321.514\" id=\"L_N2_N3_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N2_N3_0\" data-points=\"W3sieCI6NzA3Ljc2NTYyNSwieSI6MjM4fSx7IngiOjcwNy43NjU2MjUsInkiOjI3NX0seyJ4Ijo2NDUuOTgyNzMyNTk5NDMxOSwieSI6MzI0fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M325.142,86L307.409,92.167C289.675,98.333,254.209,110.667,236.475,129.5C218.742,148.333,218.742,173.667,218.742,199C218.742,224.333,218.742,249.667,218.742,267.833C218.742,286,218.742,297,218.742,302.5L218.742,308\" id=\"L_N1_N4_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N1_N4_0\" data-points=\"W3sieCI6MzI1LjE0MTkwOTk1MDY1NzksInkiOjg2fSx7IngiOjIxOC43NDIxODc1LCJ5IjoxMjN9LHsieCI6MjE4Ljc0MjE4NzUsInkiOjE5OX0seyJ4IjoyMTguNzQyMTg3NSwieSI6Mjc1fSx7IngiOjIxOC43NDIxODc1LCJ5IjozMTJ9XQ==\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M389.082,822L389.082,828.167C389.082,834.333,389.082,846.667,382.878,858.548C376.675,870.43,364.267,881.86,358.063,887.575L351.86,893.29\" id=\"L_N7_N8_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N7_N8_0\" data-points=\"W3sieCI6Mzg5LjA4MjAzMTI1LCJ5Ijo4MjJ9LHsieCI6Mzg5LjA4MjAzMTI1LCJ5Ijo4NTl9LHsieCI6MzQ4LjkxNzU1NzU2NTc4OTUsInkiOjg5Nn1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M333.476,414L347.349,420.167C361.223,426.333,388.969,438.667,414.416,450.699C439.864,462.731,463.013,474.461,474.588,480.327L486.162,486.192\" id=\"L_N4_N5_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N5_0\" data-points=\"W3sieCI6MzMzLjQ3NjM0MDU1Mzk3NzI1LCJ5Ijo0MTR9LHsieCI6NDE2LjcxNDg0Mzc1LCJ5Ijo0NTF9LHsieCI6NDg5LjczMDIyNDYwOTM3NSwieSI6NDg4fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M255.118,414L259.516,420.167C263.914,426.333,272.711,438.667,277.109,455.5C281.508,472.333,281.508,493.667,281.508,515C281.508,536.333,281.508,557.667,296.761,574.259C312.014,590.851,342.521,602.701,357.774,608.626L373.028,614.552\" id=\"L_N4_N6_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N6_0\" data-points=\"W3sieCI6MjU1LjExNzcyMDE3MDQ1NDU2LCJ5Ijo0MTR9LHsieCI6MjgxLjUwNzgxMjUsInkiOjQ1MX0seyJ4IjoyODEuNTA3ODEyNSwieSI6NTE1fSx7IngiOjI4MS41MDc4MTI1LCJ5Ijo1Nzl9LHsieCI6Mzc2Ljc1NjE2NDU1MDc4MTI1LCJ5Ijo2MTZ9XQ==\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M199.692,414L197.388,420.167C195.085,426.333,190.478,438.667,188.175,455.5C185.871,472.333,185.871,493.667,185.871,515C185.871,536.333,185.871,557.667,185.871,579C185.871,600.333,185.871,621.667,185.871,643C185.871,664.333,185.871,685.667,201.735,702.266C217.599,718.866,249.328,730.733,265.192,736.666L281.056,742.599\" id=\"L_N4_N7_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N7_0\" data-points=\"W3sieCI6MTk5LjY5MTg5NDUzMTI1LCJ5Ijo0MTR9LHsieCI6MTg1Ljg3MTA5Mzc1LCJ5Ijo0NTF9LHsieCI6MTg1Ljg3MTA5Mzc1LCJ5Ijo1MTV9LHsieCI6MTg1Ljg3MTA5Mzc1LCJ5Ijo1Nzl9LHsieCI6MTg1Ljg3MTA5Mzc1LCJ5Ijo2NDN9LHsieCI6MTg1Ljg3MTA5Mzc1LCJ5Ijo3MDd9LHsieCI6Mjg0LjgwMjczNDM3NSwieSI6NzQ0fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M446.262,670L446.262,676.167C446.262,682.333,446.262,694.667,442.023,706.467C437.784,718.268,429.307,729.536,425.068,735.17L420.829,740.804\" id=\"L_N6_N7_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N6_N7_0\" data-points=\"W3sieCI6NDQ2LjI2MTcxODc1LCJ5Ijo2NzB9LHsieCI6NDQ2LjI2MTcxODc1LCJ5Ijo3MDd9LHsieCI6NDE4LjQyNDIzOTMwOTIxMDUsInkiOjc0NH1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M596.809,402L596.809,410.167C596.809,418.333,596.809,434.667,592.054,448.49C587.299,462.313,577.79,473.625,573.036,479.282L568.281,484.938\" id=\"L_N3_N5_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N3_N5_0\" data-points=\"W3sieCI6NTk2LjgwODU5Mzc1LCJ5Ijo0MDJ9LHsieCI6NTk2LjgwODU5Mzc1LCJ5Ijo0NTF9LHsieCI6NTY1LjcwNzI3NTM5MDYyNSwieSI6NDg4fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M543.012,542L543.012,548.167C543.012,554.333,543.012,566.667,534.245,578.632C525.479,590.598,507.947,602.195,499.181,607.994L490.414,613.793\" id=\"L_N5_N6_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N5_N6_0\" data-points=\"W3sieCI6NTQzLjAxMTcxODc1LCJ5Ijo1NDJ9LHsieCI6NTQzLjAxMTcxODc1LCJ5Ijo1Nzl9LHsieCI6NDg3LjA3ODEyNSwieSI6NjE2fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/></g><g class=\"edgeLabels\"><g class=\"edgeLabel\" transform=\"translate(485.8515625, 199)\"><g class=\"label\" data-id=\"L_N1_N3_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(707.765625, 123)\"><g class=\"label\" data-id=\"L_N1_N2_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(64.9140625, 643)\"><g class=\"label\" data-id=\"L_N4_N8_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(707.765625, 275)\"><g class=\"label\" data-id=\"L_N2_N3_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(218.7421875, 199)\"><g class=\"label\" data-id=\"L_N1_N4_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(389.08203125, 859)\"><g class=\"label\" data-id=\"L_N7_N8_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(416.71484375, 451)\"><g class=\"label\" data-id=\"L_N4_N5_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(281.5078125, 515)\"><g class=\"label\" data-id=\"L_N4_N6_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(185.87109375, 579)\"><g class=\"label\" data-id=\"L_N4_N7_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(446.26171875, 707)\"><g class=\"label\" data-id=\"L_N6_N7_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(596.80859375, 451)\"><g class=\"label\" data-id=\"L_N3_N5_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(543.01171875, 579)\"><g class=\"label\" data-id=\"L_N5_N6_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g></g><g class=\"nodes\"><g class=\"node default\" id=\"flowchart-N1-0\" transform=\"translate(437.29296875, 47)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>Module 1: Introduction to Galaxy</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N3-1\" transform=\"translate(596.80859375, 363)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(1.2) Galaxy Basics for genomics</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N2-5\" transform=\"translate(707.765625, 199)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(1.1) A short introduction to Galaxy</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N4-8\" transform=\"translate(218.7421875, 363)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-51\" width=\"260\" height=\"102\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -36)\"><rect/><foreignObject width=\"200\" height=\"72\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>Module 2: Basics of Genome Sequence Analysis</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N8-9\" transform=\"translate(306.58203125, 935)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(2.4) Chloroplast genome assembly</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N7-20\" transform=\"translate(389.08203125, 783)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(2.3) An Introduction to Genome Assembly</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N5-25\" transform=\"translate(543.01171875, 515)\"><rect class=\"basic label-container\" style=\"\" x=\"-101.5859375\" y=\"-27\" width=\"203.171875\" height=\"54\"/><g class=\"label\" style=\"\" transform=\"translate(-71.5859375, -12)\"><rect/><foreignObject width=\"143.171875\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"nodeLabel\"><p>(2.1) Quality Control</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N6-29\" transform=\"translate(446.26171875, 643)\"><rect class=\"basic label-container\" style=\"\" x=\"-79.359375\" y=\"-27\" width=\"158.71875\" height=\"54\"/><g class=\"label\" style=\"\" transform=\"translate(-49.359375, -12)\"><rect/><foreignObject width=\"98.71875\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"nodeLabel\"><p>(2.2) Mapping</p></span></div></foreignObject></g></g></g></g></g></svg>"
     },
     "metadata": {},
     "output_type": "display_data",
     "jetTransient": {
      "display_id": null
     }
    }
   ],
   "execution_count": 43
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-04T12:27:49.720270Z",
     "start_time": "2025-12-04T12:27:49.711626Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
