{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QgLFG7sRFw5m"
      },
      "source": [
        "# Prototype notebook to semantically represent Learning Path with Schema.org\n",
        "\n",
        "authors: Phil Reed, Alban Gaignard, Leyla Jael Castro\n",
        "\n",
        "- Initially drafted the 20th of November 2025, as part of Bioschemas activities and ELIXIR BioHackathon Europe 2025.\n",
        "- Refined at de.NBI BioHackathon Germany 2025.\n",
        "- To be revised by the community.\n",
        "\n",
        "Any questions, please contact phil.reed@manchester.ac.uk, alban.gaignard@univ-nantes.fr\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 98,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bic7lfnooe6_",
        "outputId": "1ace6255-54f1-4613-985e-72e389702347"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "python(4598) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: rdflib in /Users/user/miniforge3/lib/python3.12/site-packages (7.5.0)\n",
            "Requirement already satisfied: pyparsing<4,>=2.1.0 in /Users/user/miniforge3/lib/python3.12/site-packages (from rdflib) (3.2.5)\n"
          ]
        }
      ],
      "source": [
        "!pip install rdflib"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_jfYP8fd1KBf"
      },
      "source": [
        "# Sample Learning path\n",
        "https://tess.elixir-europe.org/learning_paths/introduction-to-galaxy-and-sequence-analysis-6384c0ed-3546-41cf-ac30-bff8680dd96c\n",
        "\n",
        "LP structure: \n",
        "\n",
        "**Introduction to Galaxy and Sequence analysis** [syllabusSections=M1,M2]\n",
        "- **Module 1: Introduction to Galaxy** [itemListElement=11,12] [nextItem=M2]\n",
        "  - (1.1) A short introduction to Galaxy [nextItem=12]\n",
        "  - (1.2) Galaxy Basics for genomics [nextItem=21]\n",
        "- **Module 2: Basics of Genome Sequence Analysis** [itemListElement=21,22,23,24]\n",
        "  - (2.1) Quality Control [nextItem=22]\n",
        "  - (2.2) Mapping [nextItem=23]\n",
        "  - (2.3) An Introduction to Genome Assembly [nextItem=24]\n",
        "  - (2.4) Chloroplast genome assembly \n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 99,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "D4rTddCwogmS",
        "outputId": "d6e07b66-6196-46f5-a3bf-f7b70e543f01"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<Graph identifier=N606f43e385714c88a50dc59c9c5dc347 (<class 'rdflib.graph.Graph'>)>"
            ]
          },
          "execution_count": 99,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from rdflib import Graph\n",
        "\n",
        "my_learning_path = \"\"\"\n",
        "@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n",
        "@prefix owl: <http://www.w3.org/2002/07/owl#> .\n",
        "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n",
        "@prefix schema: <https://schema.org/> .\n",
        "@prefix dct: <http://purl.org/dc/terms/> .\n",
        "@prefix ex: <http://example.org/> .\n",
        "ex:GA_learning_path a schema:Course ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/LearningPath> ;\n",
        "    schema:name \"Introduction to Galaxy and Sequence analysis\" ;\n",
        "    schema:description \"This learning path aims to teach you the basics of Galaxy and analysis of sequencing data. \" ;\n",
        "    schema:provider ex:ExampleUniversity ;\n",
        "    schema:courseCode \"GSA101\" ;\n",
        "    schema:syllabusSections ex:Module_2, ex:Module_1 .\n",
        "\t\n",
        "ex:Module_1 a schema:Syllabus, schema:ListItem, schema:ItemList ;  \n",
        "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
        "    schema:name \"Module 1: Introduction to Galaxy\" ;\n",
        "    schema:teaches \"Learn how to create a workflow\" ;\n",
        "    schema:nextItem ex:Module_2 ;\n",
        "    schema:itemListElement ex:TM11, ex:TM12 .\n",
        "ex:TM11 a schema:LearningResource, schema:ListItem ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
        "    schema:name \"(1.1) A short introduction to Galaxy\" ;\n",
        "    schema:description \"What is Galaxy\" ;\n",
        "    schema:url \"https://tess.elixir-europe.org/materials/hands-on-for-a-short-introduction-to-galaxy-tutorial?lp=1%3A1\" ;\n",
        "    schema:nextItem ex:TM12 .\n",
        "ex:TM12 a schema:LearningResource, schema:ListItem ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
        "    schema:name \"(1.2) Galaxy Basics for genomics\" ;\n",
        "    schema:description \"The basic details for Galaxy genomics\" ;\n",
        "    schema:nextItem ex:TM21 .\n",
        "\n",
        "ex:Module_2 a schema:Syllabus, schema:ListItem, schema:ItemList  ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
        "    schema:name \"Module 2: Basics of Genome Sequence Analysis\" ;\n",
        "    schema:teaches \"Use genome browser to understand your data\" ;\n",
        "    schema:itemListElement ex:TM21, ex:TM22, ex:TM23, ex:TM24 .\n",
        "ex:TM21 a schema:LearningResource, schema:ListItem ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
        "    schema:name \"(2.1) Quality Control\" ;\n",
        "    schema:description \"Ensure you follow quality control practices\" ;\n",
        "    schema:nextItem ex:TM22 .\n",
        "ex:TM22 a schema:LearningResource, schema:ListItem ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
        "    schema:name \"(2.2) Mapping\" ;\n",
        "    schema:description \"Alignment and mapping\" ;\n",
        "    schema:nextItem ex:TM23 .\n",
        "ex:TM23 a schema:LearningResource, schema:ListItem ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
        "    schema:name \"(2.3) An Introduction to Genome Assembly\" ;\n",
        "    schema:description \"Description of TrainingMaterial 23\" ;\n",
        "    schema:nextItem ex:TM24 .\n",
        "ex:TM24 a schema:LearningResource, schema:ListItem ;\n",
        "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
        "    schema:name \"(2.4) Chloroplast genome assembly\" ;\n",
        "    schema:description \"Description of TrainingMaterial 24\"  .\n",
        "\"\"\"\n",
        "\n",
        "kg = Graph()\n",
        "kg.parse(data=my_learning_path, format=\"turtle\")\n",
        "kg.serialize(destination=\"lp.ttl\", format=\"turtle\")\n",
        "kg.serialize(destination=\"lp.jsonld\", format=\"json-ld\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 100,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nUcUKOMDpqkf",
        "outputId": "7ebaf0a1-1603-400f-b9d3-3d712eba24d7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "@prefix dct: <http://purl.org/dc/terms/> .\n",
            "@prefix ex: <http://example.org/> .\n",
            "@prefix schema: <https://schema.org/> .\n",
            "\n",
            "ex:GA_learning_path a schema:Course ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/LearningPath> ;\n",
            "    schema:courseCode \"GSA101\" ;\n",
            "    schema:description \"This learning path aims to teach you the basics of Galaxy and analysis of sequencing data. \" ;\n",
            "    schema:name \"Introduction to Galaxy and Sequence analysis\" ;\n",
            "    schema:provider ex:ExampleUniversity ;\n",
            "    schema:syllabusSections ex:Module_1,\n",
            "        ex:Module_2 .\n",
            "\n",
            "ex:Module_1 a schema:ItemList,\n",
            "        schema:ListItem,\n",
            "        schema:Syllabus ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
            "    schema:itemListElement ex:TM11,\n",
            "        ex:TM12 ;\n",
            "    schema:name \"Module 1: Introduction to Galaxy\" ;\n",
            "    schema:nextItem ex:Module_2 ;\n",
            "    schema:teaches \"Learn how to create a workflow\" .\n",
            "\n",
            "ex:TM11 a schema:LearningResource,\n",
            "        schema:ListItem ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
            "    schema:description \"What is Galaxy\" ;\n",
            "    schema:name \"(1.1) A short introduction to Galaxy\" ;\n",
            "    schema:nextItem ex:TM12 ;\n",
            "    schema:url \"https://tess.elixir-europe.org/materials/hands-on-for-a-short-introduction-to-galaxy-tutorial?lp=1%3A1\" .\n",
            "\n",
            "ex:Module_2 a schema:ItemList,\n",
            "        schema:ListItem,\n",
            "        schema:Syllabus ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/LearningPathModule> ;\n",
            "    schema:itemListElement ex:TM21,\n",
            "        ex:TM22,\n",
            "        ex:TM23,\n",
            "        ex:TM24 ;\n",
            "    schema:name \"Module 2: Basics of Genome Sequence Analysis\" ;\n",
            "    schema:teaches \"Use genome browser to understand your data\" .\n",
            "\n",
            "ex:TM12 a schema:LearningResource,\n",
            "        schema:ListItem ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
            "    schema:description \"The basic details for Galaxy genomics\" ;\n",
            "    schema:name \"(1.2) Galaxy Basics for genomics\" ;\n",
            "    schema:nextItem ex:TM21 .\n",
            "\n",
            "ex:TM21 a schema:LearningResource,\n",
            "        schema:ListItem ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
            "    schema:description \"Ensure you follow quality control practices\" ;\n",
            "    schema:name \"(2.1) Quality Control\" ;\n",
            "    schema:nextItem ex:TM22 .\n",
            "\n",
            "ex:TM22 a schema:LearningResource,\n",
            "        schema:ListItem ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
            "    schema:description \"Alignment and mapping\" ;\n",
            "    schema:name \"(2.2) Mapping\" ;\n",
            "    schema:nextItem ex:TM23 .\n",
            "\n",
            "ex:TM23 a schema:LearningResource,\n",
            "        schema:ListItem ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
            "    schema:description \"Description of TrainingMaterial 23\" ;\n",
            "    schema:name \"(2.3) An Introduction to Genome Assembly\" ;\n",
            "    schema:nextItem ex:TM24 .\n",
            "\n",
            "ex:TM24 a schema:LearningResource,\n",
            "        schema:ListItem ;\n",
            "    dct:conformsTo <https://bioschemas.org/profiles/TrainingMaterial> ;\n",
            "    schema:description \"Description of TrainingMaterial 24\" ;\n",
            "    schema:name \"(2.4) Chloroplast genome assembly\" .\n",
            "\n",
            "\n",
            "63\n"
          ]
        }
      ],
      "source": [
        "print(kg.serialize(format=\"turtle\"))\n",
        "print(len(kg))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 101,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pzIQoZAt3V0g",
        "outputId": "1da6d4ed-3b16-45f5-f8d0-9318a37833f7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[\n",
            "  {\n",
            "    \"@id\": \"http://example.org/GA_learning_path\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/Course\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/LearningPath\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/courseCode\": [\n",
            "      {\n",
            "        \"@value\": \"GSA101\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/description\": [\n",
            "      {\n",
            "        \"@value\": \"This learning path aims to teach you the basics of Galaxy and analysis of sequencing data. \"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"Introduction to Galaxy and Sequence analysis\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/provider\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/ExampleUniversity\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/syllabusSections\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/Module_2\"\n",
            "      },\n",
            "      {\n",
            "        \"@id\": \"http://example.org/Module_1\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/TM23\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/LearningResource\",\n",
            "      \"https://schema.org/ListItem\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/description\": [\n",
            "      {\n",
            "        \"@value\": \"Description of TrainingMaterial 23\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"(2.3) An Introduction to Genome Assembly\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/nextItem\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM24\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/Module_1\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/Syllabus\",\n",
            "      \"https://schema.org/ListItem\",\n",
            "      \"https://schema.org/ItemList\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/LearningPathModule\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/itemListElement\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM11\"\n",
            "      },\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM12\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"Module 1: Introduction to Galaxy\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/nextItem\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/Module_2\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/teaches\": [\n",
            "      {\n",
            "        \"@value\": \"Learn how to create a workflow\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/TM11\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/LearningResource\",\n",
            "      \"https://schema.org/ListItem\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/description\": [\n",
            "      {\n",
            "        \"@value\": \"What is Galaxy\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"(1.1) A short introduction to Galaxy\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/nextItem\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM12\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/url\": [\n",
            "      {\n",
            "        \"@value\": \"https://tess.elixir-europe.org/materials/hands-on-for-a-short-introduction-to-galaxy-tutorial?lp=1%3A1\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/TM12\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/LearningResource\",\n",
            "      \"https://schema.org/ListItem\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/description\": [\n",
            "      {\n",
            "        \"@value\": \"The basic details for Galaxy genomics\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"(1.2) Galaxy Basics for genomics\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/nextItem\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM21\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/Module_2\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/Syllabus\",\n",
            "      \"https://schema.org/ListItem\",\n",
            "      \"https://schema.org/ItemList\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/LearningPathModule\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/itemListElement\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM21\"\n",
            "      },\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM22\"\n",
            "      },\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM23\"\n",
            "      },\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM24\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"Module 2: Basics of Genome Sequence Analysis\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/teaches\": [\n",
            "      {\n",
            "        \"@value\": \"Use genome browser to understand your data\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/TM22\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/LearningResource\",\n",
            "      \"https://schema.org/ListItem\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/description\": [\n",
            "      {\n",
            "        \"@value\": \"Alignment and mapping\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"(2.2) Mapping\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/nextItem\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM23\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/TM24\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/LearningResource\",\n",
            "      \"https://schema.org/ListItem\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/description\": [\n",
            "      {\n",
            "        \"@value\": \"Description of TrainingMaterial 24\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"(2.4) Chloroplast genome assembly\"\n",
            "      }\n",
            "    ]\n",
            "  },\n",
            "  {\n",
            "    \"@id\": \"http://example.org/TM21\",\n",
            "    \"@type\": [\n",
            "      \"https://schema.org/LearningResource\",\n",
            "      \"https://schema.org/ListItem\"\n",
            "    ],\n",
            "    \"http://purl.org/dc/terms/conformsTo\": [\n",
            "      {\n",
            "        \"@id\": \"https://bioschemas.org/profiles/TrainingMaterial\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/description\": [\n",
            "      {\n",
            "        \"@value\": \"Ensure you follow quality control practices\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/name\": [\n",
            "      {\n",
            "        \"@value\": \"(2.1) Quality Control\"\n",
            "      }\n",
            "    ],\n",
            "    \"https://schema.org/nextItem\": [\n",
            "      {\n",
            "        \"@id\": \"http://example.org/TM22\"\n",
            "      }\n",
            "    ]\n",
            "  }\n",
            "]\n"
          ]
        }
      ],
      "source": [
        "print(kg.serialize(format=\"json-ld\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 102,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "P2wsCCjvqKzV",
        "outputId": "e9941f9a-2ea7-45e9-d92b-fce817303d34"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "If you are interested in (2.3) An Introduction to Genome Assembly\n",
            "You need to perform (2.2) Mapping as part of the module Module 2: Basics of Genome Sequence Analysis\n",
            "You need to perform (2.1) Quality Control as part of the module Module 2: Basics of Genome Sequence Analysis\n",
            "You need to perform (1.2) Galaxy Basics for genomics as part of the module Module 1: Introduction to Galaxy\n",
            "You need to perform (1.1) A short introduction to Galaxy as part of the module Module 1: Introduction to Galaxy\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "<rdflib.plugins.sparql.processor.SPARQLResult at 0x10a3ec980>"
            ]
          },
          "execution_count": 102,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# I'm interested in a specific training material, what should be the prerequisites ?\n",
        "\n",
        "tm_of_interest = \"(2.3) An Introduction to Genome Assembly\"\n",
        "\n",
        "def query_learning_path():\n",
        "  query = f\"\"\"\n",
        "  SELECT * WHERE {{\n",
        "    ?s schema:name \"{tm_of_interest}\" .\n",
        "    ?prereq schema:nextItem+ ?s .\n",
        "    ?prereq schema:name ?prereq_name .\n",
        "    ?a_module schema:itemListElement ?prereq .\n",
        "    ?a_module schema:name ?module_name .\n",
        "  }}\n",
        "  \"\"\"\n",
        "  # Run SPARQL query\n",
        "  res_kg = kg.query(query)\n",
        "  print(f'If you are interested in {tm_of_interest}')\n",
        "  for r in res_kg:\n",
        "    print(f'You need to perform {r[\"prereq_name\"]} as part of the module {r[\"module_name\"]}')\n",
        "  return res_kg\n",
        "\n",
        "query_learning_path()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9ff01bd3"
      },
      "source": [
        "## What we did until now\n",
        "\n",
        "1.  **RDF Graph Initialization**: We loaded a learning path described in Turtle format into an RDFLib graph (`kg`). We extensively used Schema.org to model the LearningPath Syllabus, ItemList and ListItem types, as well as Course and LearningResource types.\n",
        "2.  **Prerequisite Query**: We demonstrated how to query the graph using SPARQL to find prerequisites for a specific learning material.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VBoX4sRKEx1j"
      },
      "source": [
        "# Visualization of the Schema.org metadata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 103,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0fd2540c",
        "outputId": "4d4eee93-c3f5-47c8-ae1d-390bd1969867"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Extracted entity names:\n",
            "  http://example.org/GA_learning_path: Introduction to Galaxy and Sequence analysis\n",
            "  http://example.org/Module_1: Module 1: Introduction to Galaxy\n",
            "  http://example.org/TM11: (1.1) A short introduction to Galaxy\n",
            "  http://example.org/TM12: (1.2) Galaxy Basics for genomics\n",
            "  http://example.org/Module_2: Module 2: Basics of Genome Sequence Analysis\n",
            "  http://example.org/TM21: (2.1) Quality Control\n",
            "  http://example.org/TM22: (2.2) Mapping\n",
            "  http://example.org/TM23: (2.3) An Introduction to Genome Assembly\n",
            "  http://example.org/TM24: (2.4) Chloroplast genome assembly\n",
            "\n",
            "Extracted raw relationships:\n",
            "  http://example.org/TM11 -- https://schema.org/nextItem --> http://example.org/TM12\n",
            "  http://example.org/TM23 -- https://schema.org/nextItem --> http://example.org/TM24\n",
            "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM23\n",
            "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM24\n",
            "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM22\n",
            "  http://example.org/TM22 -- https://schema.org/nextItem --> http://example.org/TM23\n",
            "  http://example.org/TM21 -- https://schema.org/nextItem --> http://example.org/TM22\n",
            "  http://example.org/Module_2 -- https://schema.org/itemListElement --> http://example.org/TM21\n",
            "  http://example.org/Module_1 -- https://schema.org/itemListElement --> http://example.org/TM12\n",
            "  http://example.org/Module_1 -- https://schema.org/itemListElement --> http://example.org/TM11\n",
            "  http://example.org/TM12 -- https://schema.org/nextItem --> http://example.org/TM21\n",
            "  http://example.org/Module_1 -- https://schema.org/nextItem --> http://example.org/Module_2\n",
            "\n",
            "Transformed relationships for Mermaid diagram:\n",
            "  (1.1) A short introduction to Galaxy -- nextItem --> (1.2) Galaxy Basics for genomics\n",
            "  (2.3) An Introduction to Genome Assembly -- nextItem --> (2.4) Chloroplast genome assembly\n",
            "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.3) An Introduction to Genome Assembly\n",
            "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.4) Chloroplast genome assembly\n",
            "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.2) Mapping\n",
            "  (2.2) Mapping -- nextItem --> (2.3) An Introduction to Genome Assembly\n",
            "  (2.1) Quality Control -- nextItem --> (2.2) Mapping\n",
            "  Module 2: Basics of Genome Sequence Analysis -- itemListElement --> (2.1) Quality Control\n",
            "  Module 1: Introduction to Galaxy -- itemListElement --> (1.2) Galaxy Basics for genomics\n",
            "  Module 1: Introduction to Galaxy -- itemListElement --> (1.1) A short introduction to Galaxy\n",
            "  (1.2) Galaxy Basics for genomics -- nextItem --> (2.1) Quality Control\n",
            "  Module 1: Introduction to Galaxy -- nextItem --> Module 2: Basics of Genome Sequence Analysis\n",
            "Relationship extraction complete. Results stored in all_entity_names and all_mermaid_edges.\n"
          ]
        }
      ],
      "source": [
        "from rdflib import Graph, Literal, URIRef\n",
        "from rdflib.namespace import RDF, RDFS, OWL, XSD\n",
        "\n",
        "def extract_relationships_and_names(graph):\n",
        "    # Define prefixes\n",
        "    schema = URIRef(\"https://schema.org/\")\n",
        "    ex = URIRef(\"http://example.org/\")\n",
        "\n",
        "    # 1. Query for all entities with a schema:name and store their names\n",
        "    entity_names = {}\n",
        "    name_query = \"\"\"\n",
        "    PREFIX schema: <https://schema.org/>\n",
        "    SELECT ?entity ?name\n",
        "    WHERE {\n",
        "        ?entity schema:name ?name .\n",
        "    }\n",
        "    \"\"\"\n",
        "    for row in graph.query(name_query):\n",
        "        entity_names[str(row.entity)] = str(row.name)\n",
        "\n",
        "    print(\"Extracted entity names:\")\n",
        "    for uri, name in entity_names.items():\n",
        "        print(f\"  {uri}: {name}\")\n",
        "\n",
        "    # 2. Query for sequential relationships\n",
        "    relationships_query = \"\"\"\n",
        "    PREFIX schema: <https://schema.org/>\n",
        "    SELECT ?s ?p ?o\n",
        "    WHERE {\n",
        "        ?s ?p ?o .\n",
        "        FILTER (?p = schema:nextItem || ?p = schema:itemListElement || ?p = schema:step)\n",
        "    }\n",
        "    \"\"\"\n",
        "    raw_relationships = []\n",
        "    for row in graph.query(relationships_query):\n",
        "        raw_relationships.append((str(row.s), str(row.p), str(row.o)))\n",
        "\n",
        "    print(\"\\nExtracted raw relationships:\")\n",
        "    for rel in raw_relationships:\n",
        "        print(f\"  {rel[0]} -- {rel[1]} --> {rel[2]}\")\n",
        "\n",
        "    # 3. Transform relationships to use names\n",
        "    mermaid_edges = []\n",
        "    for s_uri, p_uri, o_uri in raw_relationships:\n",
        "        s_name = entity_names.get(s_uri, s_uri.split('/')[-1]) # Use URI fragment if name not found\n",
        "        o_name = entity_names.get(o_uri, o_uri.split('/')[-1]) # Use URI fragment if name not found\n",
        "        predicate_type = p_uri.split('/')[-1] # Get predicate name (e.g., nextItem)\n",
        "        mermaid_edges.append((s_name, predicate_type, o_name))\n",
        "\n",
        "    print(\"\\nTransformed relationships for Mermaid diagram:\")\n",
        "    for edge in mermaid_edges:\n",
        "        print(f\"  {edge[0]} -- {edge[1]} --> {edge[2]}\")\n",
        "\n",
        "    return entity_names, mermaid_edges\n",
        "\n",
        "# Call the function with the existing kg graph\n",
        "all_entity_names, all_mermaid_edges = extract_relationships_and_names(kg)\n",
        "\n",
        "# Assign to the global query_learning_path function as well if needed later\n",
        "def query_learning_path():\n",
        "    return all_entity_names, all_mermaid_edges\n",
        "\n",
        "print(\"Relationship extraction complete. Results stored in all_entity_names and all_mermaid_edges.\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 106,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 105
        },
        "id": "9a40009b",
        "outputId": "42b9dcb4-3ad0-4a21-f442-de4850033ba4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "graph TD\n",
            "N2[\"(1.1) A short introduction to Galaxy\"]\n",
            "N3[\"(1.2) Galaxy Basics for genomics\"]\n",
            "N2 -- nextItem --> N3\n",
            "N7[\"(2.3) An Introduction to Genome Assembly\"]\n",
            "N8[\"(2.4) Chloroplast genome assembly\"]\n",
            "N7 -- nextItem --> N8\n",
            "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
            "N7[\"(2.3) An Introduction to Genome Assembly\"]\n",
            "N4 -- itemListElement --> N7\n",
            "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
            "N8[\"(2.4) Chloroplast genome assembly\"]\n",
            "N4 -- itemListElement --> N8\n",
            "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
            "N6[\"(2.2) Mapping\"]\n",
            "N4 -- itemListElement --> N6\n",
            "N6[\"(2.2) Mapping\"]\n",
            "N7[\"(2.3) An Introduction to Genome Assembly\"]\n",
            "N6 -- nextItem --> N7\n",
            "N5[\"(2.1) Quality Control\"]\n",
            "N6[\"(2.2) Mapping\"]\n",
            "N5 -- nextItem --> N6\n",
            "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
            "N5[\"(2.1) Quality Control\"]\n",
            "N4 -- itemListElement --> N5\n",
            "N1[\"Module 1: Introduction to Galaxy\"]\n",
            "N3[\"(1.2) Galaxy Basics for genomics\"]\n",
            "N1 -- itemListElement --> N3\n",
            "N1[\"Module 1: Introduction to Galaxy\"]\n",
            "N2[\"(1.1) A short introduction to Galaxy\"]\n",
            "N1 -- itemListElement --> N2\n",
            "N3[\"(1.2) Galaxy Basics for genomics\"]\n",
            "N5[\"(2.1) Quality Control\"]\n",
            "N3 -- nextItem --> N5\n",
            "N1[\"Module 1: Introduction to Galaxy\"]\n",
            "N4[\"Module 2: Basics of Genome Sequence Analysis\"]\n",
            "N1 -- nextItem --> N4\n"
          ]
        }
      ],
      "source": [
        "mermaid_lines = []\n",
        "mermaid_lines.append('graph TD')\n",
        "\n",
        "for s_name, p_type, o_name in all_mermaid_edges:\n",
        "    # Sanitize names for Mermaid if they contain special characters or spaces\n",
        "    # Mermaid node IDs cannot contain spaces or special characters directly\n",
        "    # We'll use a simple approach: if a name has spaces, enclose it in quotes.\n",
        "    # For simplicity, we'll map original names to safe IDs and use names for labels\n",
        "\n",
        "    # Create a unique ID for each node if it's not already in the entity_names (which are already unique names)\n",
        "    # For this exercise, we will assume the names are unique enough for node IDs or simply use them as both ID and label.\n",
        "    # Mermaid allows IDs with spaces if quoted, but it's cleaner to have simple IDs.\n",
        "    # However, given the prompt directly asks for f'{source_name} -- {predicate_type} --> {target_name}',\n",
        "    # we will use the names as they are, and Mermaid should handle simple spaces correctly in labels.\n",
        "\n",
        "    # Let's ensure node labels are quoted if they contain spaces to be robust\n",
        "    # Mermaid syntax: A[\"Node Name\"] or A --> B[\"Another Node\"]\n",
        "    # For simplicity, let's use the names directly as node IDs and labels.\n",
        "    # Mermaid handles spaces in node labels gracefully if they are quoted, but for connections\n",
        "    # the IDs should be simple. Let's make sure our approach is compatible.\n",
        "\n",
        "    # A common Mermaid practice is to have a simple ID for the node and then a label\n",
        "    # like: ID[Label]. If we use the name as ID, it should not have spaces.\n",
        "    # Given the previous step produced names like 'Module 1', we should create a mapping for node IDs.\n",
        "\n",
        "    # Let's refine how we generate the Mermaid lines to ensure valid syntax\n",
        "    # Create a mapping from clean name to a unique, mermaid-safe ID\n",
        "    # This mapping is important to handle cases where names might have characters that Mermaid IDs don't like\n",
        "    node_id_map = {}\n",
        "    counter = 0\n",
        "    for uri_name in all_entity_names.values():\n",
        "        clean_name = uri_name.replace(' ', '_').replace('-', '_').replace(':', '_').replace('.', '_').strip()\n",
        "        if clean_name not in node_id_map.values():\n",
        "            node_id_map[uri_name] = f'N{counter}'\n",
        "            counter += 1\n",
        "\n",
        "    # Create mermaid lines with unique IDs and original names as labels\n",
        "    mermaid_lines = ['graph TD']\n",
        "    for s_name, p_type, o_name in all_mermaid_edges:\n",
        "        source_id = node_id_map.get(s_name, s_name.replace(' ', '_').replace('-', '_').replace(':', '_').replace('.', '_').strip())\n",
        "        target_id = node_id_map.get(o_name, o_name.replace(' ', '_').replace('-', '_').replace(':', '_').replace('.', '_').strip())\n",
        "\n",
        "        # Add node definitions with labels if they contain spaces\n",
        "        if ' ' in s_name:\n",
        "            mermaid_lines.append(f'{source_id}[\\\"{s_name}\\\"]')\n",
        "        if ' ' in o_name:\n",
        "            mermaid_lines.append(f'{target_id}[\\\"{o_name}\\\"]')\n",
        "\n",
        "        mermaid_lines.append(f'{source_id} -- {p_type} --> {target_id}')\n",
        "\n",
        "mermaid_syntax = '\\n'.join(mermaid_lines)\n",
        "\n",
        "# save mermaid file to disk\n",
        "with open(\"lp.mmd\", \"w\") as mermaid_file:\n",
        "    print(mermaid_syntax, file=mermaid_file)\n",
        "\n",
        "# display the markdown mermaid graph as output of this cell\n",
        "print(mermaid_syntax)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 105,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 203
        },
        "id": "COeGzKUsAhZ2",
        "outputId": "c8203d22-cf3f-457c-8ef5-f3132f636033"
      },
      "outputs": [
        {
          "data": {
            "image/svg+xml": [
              "<svg id=\"mermaidInkSvg\" width=\"100%\" xmlns=\"http://www.w3.org/2000/svg\" class=\"flowchart\" style=\"max-width: 819.859375px;\" viewBox=\"0 0 819.859375 982\" role=\"graphics-document document\" aria-roledescription=\"flowchart-v2\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><style xmlns=\"http://www.w3.org/1999/xhtml\">@import url(\"https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.7.2/css/all.min.css\");</style><style>#mermaidInkSvg{font-family:\"trebuchet ms\",verdana,arial,sans-serif;font-size:16px;fill:#333;}@keyframes edge-animation-frame{from{stroke-dashoffset:0;}}@keyframes dash{to{stroke-dashoffset:0;}}#mermaidInkSvg .edge-animation-slow{stroke-dasharray:9,5!important;stroke-dashoffset:900;animation:dash 50s linear infinite;stroke-linecap:round;}#mermaidInkSvg .edge-animation-fast{stroke-dasharray:9,5!important;stroke-dashoffset:900;animation:dash 20s linear infinite;stroke-linecap:round;}#mermaidInkSvg .error-icon{fill:#552222;}#mermaidInkSvg .error-text{fill:#552222;stroke:#552222;}#mermaidInkSvg .edge-thickness-normal{stroke-width:1px;}#mermaidInkSvg .edge-thickness-thick{stroke-width:3.5px;}#mermaidInkSvg .edge-pattern-solid{stroke-dasharray:0;}#mermaidInkSvg .edge-thickness-invisible{stroke-width:0;fill:none;}#mermaidInkSvg .edge-pattern-dashed{stroke-dasharray:3;}#mermaidInkSvg .edge-pattern-dotted{stroke-dasharray:2;}#mermaidInkSvg .marker{fill:#333333;stroke:#333333;}#mermaidInkSvg .marker.cross{stroke:#333333;}#mermaidInkSvg svg{font-family:\"trebuchet ms\",verdana,arial,sans-serif;font-size:16px;}#mermaidInkSvg p{margin:0;}#mermaidInkSvg .label{font-family:\"trebuchet ms\",verdana,arial,sans-serif;color:#333;}#mermaidInkSvg .cluster-label text{fill:#333;}#mermaidInkSvg .cluster-label span{color:#333;}#mermaidInkSvg .cluster-label span p{background-color:transparent;}#mermaidInkSvg .label text,#mermaidInkSvg span{fill:#333;color:#333;}#mermaidInkSvg .node rect,#mermaidInkSvg .node circle,#mermaidInkSvg .node ellipse,#mermaidInkSvg .node polygon,#mermaidInkSvg .node path{fill:#ECECFF;stroke:#9370DB;stroke-width:1px;}#mermaidInkSvg .rough-node .label text,#mermaidInkSvg .node .label text,#mermaidInkSvg .image-shape .label,#mermaidInkSvg .icon-shape .label{text-anchor:middle;}#mermaidInkSvg .node .katex path{fill:#000;stroke:#000;stroke-width:1px;}#mermaidInkSvg .rough-node .label,#mermaidInkSvg .node .label,#mermaidInkSvg .image-shape .label,#mermaidInkSvg .icon-shape .label{text-align:center;}#mermaidInkSvg .node.clickable{cursor:pointer;}#mermaidInkSvg .root .anchor path{fill:#333333!important;stroke-width:0;stroke:#333333;}#mermaidInkSvg .arrowheadPath{fill:#333333;}#mermaidInkSvg .edgePath .path{stroke:#333333;stroke-width:2.0px;}#mermaidInkSvg .flowchart-link{stroke:#333333;fill:none;}#mermaidInkSvg .edgeLabel{background-color:rgba(232,232,232, 0.8);text-align:center;}#mermaidInkSvg .edgeLabel p{background-color:rgba(232,232,232, 0.8);}#mermaidInkSvg .edgeLabel rect{opacity:0.5;background-color:rgba(232,232,232, 0.8);fill:rgba(232,232,232, 0.8);}#mermaidInkSvg .labelBkg{background-color:rgba(232, 232, 232, 0.5);}#mermaidInkSvg .cluster rect{fill:#ffffde;stroke:#aaaa33;stroke-width:1px;}#mermaidInkSvg .cluster text{fill:#333;}#mermaidInkSvg .cluster span{color:#333;}#mermaidInkSvg div.mermaidTooltip{position:absolute;text-align:center;max-width:200px;padding:2px;font-family:\"trebuchet ms\",verdana,arial,sans-serif;font-size:12px;background:hsl(80, 100%, 96.2745098039%);border:1px solid #aaaa33;border-radius:2px;pointer-events:none;z-index:100;}#mermaidInkSvg .flowchartTitleText{text-anchor:middle;font-size:18px;fill:#333;}#mermaidInkSvg rect.text{fill:none;stroke-width:0;}#mermaidInkSvg .icon-shape,#mermaidInkSvg .image-shape{background-color:rgba(232,232,232, 0.8);text-align:center;}#mermaidInkSvg .icon-shape p,#mermaidInkSvg .image-shape p{background-color:rgba(232,232,232, 0.8);padding:2px;}#mermaidInkSvg .icon-shape rect,#mermaidInkSvg .image-shape rect{opacity:0.5;background-color:rgba(232,232,232, 0.8);fill:rgba(232,232,232, 0.8);}#mermaidInkSvg .label-icon{display:inline-block;height:1em;overflow:visible;vertical-align:-0.125em;}#mermaidInkSvg .node .label-icon path{fill:currentColor;stroke:revert;stroke-width:revert;}#mermaidInkSvg :root{--mermaid-font-family:\"trebuchet ms\",verdana,arial,sans-serif;}</style><g><marker id=\"mermaidInkSvg_flowchart-v2-pointEnd\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"5\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"8\" markerHeight=\"8\" orient=\"auto\"><path d=\"M 0 0 L 10 5 L 0 10 z\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-pointStart\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"4.5\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"8\" markerHeight=\"8\" orient=\"auto\"><path d=\"M 0 5 L 10 10 L 10 0 z\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-circleEnd\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"11\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><circle cx=\"5\" cy=\"5\" r=\"5\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-circleStart\" class=\"marker flowchart-v2\" viewBox=\"0 0 10 10\" refX=\"-1\" refY=\"5\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><circle cx=\"5\" cy=\"5\" r=\"5\" class=\"arrowMarkerPath\" style=\"stroke-width: 1; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-crossEnd\" class=\"marker cross flowchart-v2\" viewBox=\"0 0 11 11\" refX=\"12\" refY=\"5.2\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><path d=\"M 1,1 l 9,9 M 10,1 l -9,9\" class=\"arrowMarkerPath\" style=\"stroke-width: 2; stroke-dasharray: 1, 0;\"/></marker><marker id=\"mermaidInkSvg_flowchart-v2-crossStart\" class=\"marker cross flowchart-v2\" viewBox=\"0 0 11 11\" refX=\"-1\" refY=\"5.2\" markerUnits=\"userSpaceOnUse\" markerWidth=\"11\" markerHeight=\"11\" orient=\"auto\"><path d=\"M 1,1 l 9,9 M 10,1 l -9,9\" class=\"arrowMarkerPath\" style=\"stroke-width: 2; stroke-dasharray: 1, 0;\"/></marker><g class=\"root\"><g class=\"clusters\"/><g class=\"edgePaths\"><path d=\"M533.031,238L533.031,244.167C533.031,250.333,533.031,262.667,535.581,276.368C538.13,290.07,543.229,305.141,545.778,312.676L548.328,320.211\" id=\"L_N2_N3_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N2_N3_0\" data-points=\"W3sieCI6NTMzLjAzMTI1LCJ5IjoyMzh9LHsieCI6NTMzLjAzMTI1LCJ5IjoyNzV9LHsieCI6NTQ5LjYwOTY0MTMzNTIyNzMsInkiOjMyNH1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M141.828,822L141.828,828.167C141.828,834.333,141.828,846.667,148.278,858.557C154.728,870.448,167.628,881.897,174.078,887.621L180.528,893.345\" id=\"L_N7_N8_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N7_N8_0\" data-points=\"W3sieCI6MTQxLjgyODEyNSwieSI6ODIyfSx7IngiOjE0MS44MjgxMjUsInkiOjg1OX0seyJ4IjoxODMuNTE5Njg1NDQ0MDc4OTYsInkiOjg5Nn1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M121.294,414L111.898,420.167C102.501,426.333,83.708,438.667,74.311,455.5C64.914,472.333,64.914,493.667,64.914,515C64.914,536.333,64.914,557.667,64.914,579C64.914,600.333,64.914,621.667,64.914,643C64.914,664.333,64.914,685.667,70.681,702.031C76.447,718.396,87.981,729.792,93.747,735.49L99.514,741.189\" id=\"L_N4_N7_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N7_0\" data-points=\"W3sieCI6MTIxLjI5NDM4OTIwNDU0NTQ1LCJ5Ijo0MTR9LHsieCI6NjQuOTE0MDYyNSwieSI6NDUxfSx7IngiOjY0LjkxNDA2MjUsInkiOjUxNX0seyJ4Ijo2NC45MTQwNjI1LCJ5Ijo1Nzl9LHsieCI6NjQuOTE0MDYyNSwieSI6NjQzfSx7IngiOjY0LjkxNDA2MjUsInkiOjcwN30seyJ4IjoxMDIuMzU5MDY2NjExODQyMTEsInkiOjc0NH1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M329.008,392.397L372.201,402.164C415.393,411.931,501.779,431.466,544.971,451.899C588.164,472.333,588.164,493.667,588.164,515C588.164,536.333,588.164,557.667,588.164,579C588.164,600.333,588.164,621.667,588.164,643C588.164,664.333,588.164,685.667,588.164,709C588.164,732.333,588.164,757.667,588.164,783C588.164,808.333,588.164,833.667,550.367,854.297C512.569,874.928,436.974,890.856,399.176,898.82L361.379,906.784\" id=\"L_N4_N8_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N8_0\" data-points=\"W3sieCI6MzI5LjAwNzgxMjUsInkiOjM5Mi4zOTY5MzI0NjYwNzI0fSx7IngiOjU4OC4xNjQwNjI1LCJ5Ijo0NTF9LHsieCI6NTg4LjE2NDA2MjUsInkiOjUxNX0seyJ4Ijo1ODguMTY0MDYyNSwieSI6NTc5fSx7IngiOjU4OC4xNjQwNjI1LCJ5Ijo2NDN9LHsieCI6NTg4LjE2NDA2MjUsInkiOjcwN30seyJ4Ijo1ODguMTY0MDYyNSwieSI6NzgzfSx7IngiOjU4OC4xNjQwNjI1LCJ5Ijo4NTl9LHsieCI6MzU3LjQ2NDg0Mzc1LCJ5Ijo5MDcuNjA4NzU2ODYzMjk3Mn1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M176.72,414L174.025,420.167C171.33,426.333,165.941,438.667,163.246,455.5C160.551,472.333,160.551,493.667,160.551,515C160.551,536.333,160.551,557.667,166.933,574.055C173.315,590.443,186.079,601.887,192.461,607.608L198.843,613.33\" id=\"L_N4_N6_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N6_0\" data-points=\"W3sieCI6MTc2LjcyMDIxNDg0Mzc1LCJ5Ijo0MTR9LHsieCI6MTYwLjU1MDc4MTI1LCJ5Ijo0NTF9LHsieCI6MTYwLjU1MDc4MTI1LCJ5Ijo1MTV9LHsieCI6MTYwLjU1MDc4MTI1LCJ5Ijo1Nzl9LHsieCI6MjAxLjgyMTIyODAyNzM0Mzc1LCJ5Ijo2MTZ9XQ==\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M231.938,670L231.938,676.167C231.938,682.333,231.938,694.667,225.136,706.57C218.334,718.474,204.73,729.947,197.928,735.684L191.126,741.421\" id=\"L_N6_N7_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N6_N7_0\" data-points=\"W3sieCI6MjMxLjkzNzUsInkiOjY3MH0seyJ4IjoyMzEuOTM3NSwieSI6NzA3fSx7IngiOjE4OC4wNjg0NjIxNzEwNTI2MywieSI6NzQ0fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M429.891,542L429.891,548.167C429.891,554.333,429.891,566.667,410.759,579.019C391.628,591.371,353.365,603.741,334.234,609.927L315.103,616.112\" id=\"L_N5_N6_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N5_N6_0\" data-points=\"W3sieCI6NDI5Ljg5MDYyNSwieSI6NTQyfSx7IngiOjQyOS44OTA2MjUsInkiOjU3OX0seyJ4IjozMTEuMjk2ODc1LCJ5Ijo2MTcuMzQyNDEwNjA4NTcyMX1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M255.079,414L261.859,420.167C268.638,426.333,282.198,438.667,301.301,450.713C320.403,462.759,345.048,474.518,357.371,480.398L369.693,486.277\" id=\"L_N4_N5_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N4_N5_0\" data-points=\"W3sieCI6MjU1LjA3ODgzNTIyNzI3MjcyLCJ5Ijo0MTR9LHsieCI6Mjk1Ljc1NzgxMjUsInkiOjQ1MX0seyJ4IjozNzMuMzAzMzQ0NzI2NTYyNSwieSI6NDg4fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M581.848,79.597L610.697,86.831C639.547,94.065,697.246,108.532,726.096,128.433C754.945,148.333,754.945,173.667,754.945,199C754.945,224.333,754.945,249.667,737.72,270.222C720.495,290.778,686.045,306.556,668.82,314.445L651.595,322.334\" id=\"L_N1_N3_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N1_N3_0\" data-points=\"W3sieCI6NTgxLjg0NzY1NjI1LCJ5Ijo3OS41OTY3NTQ4NjE5MDc2NH0seyJ4Ijo3NTQuOTQ1MzEyNSwieSI6MTIzfSx7IngiOjc1NC45NDUzMTI1LCJ5IjoxOTl9LHsieCI6NzU0Ljk0NTMxMjUsInkiOjI3NX0seyJ4Ijo2NDcuOTU3OTE5MDM0MDkwOSwieSI6MzI0fV0=\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M493.508,86L500.095,92.167C506.682,98.333,519.857,110.667,526.444,122.333C533.031,134,533.031,145,533.031,150.5L533.031,156\" id=\"L_N1_N2_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N1_N2_0\" data-points=\"W3sieCI6NDkzLjUwNzY1ODMwNTkyMTA0LCJ5Ijo4Nn0seyJ4Ijo1MzMuMDMxMjUsInkiOjEyM30seyJ4Ijo1MzMuMDMxMjUsInkiOjE2MH1d\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M551.583,402L549.233,410.167C546.884,418.333,542.184,434.667,530.04,448.659C517.896,462.652,498.308,474.303,488.514,480.129L478.72,485.955\" id=\"L_N3_N5_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N3_N5_0\" data-points=\"W3sieCI6NTUxLjU4MzE4NTM2OTMxODEsInkiOjQwMn0seyJ4Ijo1MzcuNDg0Mzc1LCJ5Ijo0NTF9LHsieCI6NDc1LjI4MTczODI4MTI1LCJ5Ijo0ODh9XQ==\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/><path d=\"M322.101,86L301.585,92.167C281.07,98.333,240.039,110.667,219.523,129.5C199.008,148.333,199.008,173.667,199.008,199C199.008,224.333,199.008,249.667,199.008,267.833C199.008,286,199.008,297,199.008,302.5L199.008,308\" id=\"L_N1_N4_0\" class=\"edge-thickness-normal edge-pattern-solid edge-thickness-normal edge-pattern-solid flowchart-link\" style=\";\" data-edge=\"true\" data-et=\"edge\" data-id=\"L_N1_N4_0\" data-points=\"W3sieCI6MzIyLjEwMDg5NDMyNTY1NzksInkiOjg2fSx7IngiOjE5OS4wMDc4MTI1LCJ5IjoxMjN9LHsieCI6MTk5LjAwNzgxMjUsInkiOjE5OX0seyJ4IjoxOTkuMDA3ODEyNSwieSI6Mjc1fSx7IngiOjE5OS4wMDc4MTI1LCJ5IjozMTJ9XQ==\" marker-end=\"url(#mermaidInkSvg_flowchart-v2-pointEnd)\"/></g><g class=\"edgeLabels\"><g class=\"edgeLabel\" transform=\"translate(533.03125, 275)\"><g class=\"label\" data-id=\"L_N2_N3_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(141.828125, 859)\"><g class=\"label\" data-id=\"L_N7_N8_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(64.9140625, 579)\"><g class=\"label\" data-id=\"L_N4_N7_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(588.1640625, 643)\"><g class=\"label\" data-id=\"L_N4_N8_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(160.55078125, 515)\"><g class=\"label\" data-id=\"L_N4_N6_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(231.9375, 707)\"><g class=\"label\" data-id=\"L_N6_N7_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(429.890625, 579)\"><g class=\"label\" data-id=\"L_N5_N6_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(295.7578125, 451)\"><g class=\"label\" data-id=\"L_N4_N5_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(754.9453125, 199)\"><g class=\"label\" data-id=\"L_N1_N3_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(533.03125, 123)\"><g class=\"label\" data-id=\"L_N1_N2_0\" transform=\"translate(-56.9140625, -12)\"><foreignObject width=\"113.828125\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>itemListElement</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(528.2938, 456.46683)\"><g class=\"label\" data-id=\"L_N3_N5_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g><g class=\"edgeLabel\" transform=\"translate(199.0078125, 199)\"><g class=\"label\" data-id=\"L_N1_N4_0\" transform=\"translate(-30.6796875, -12)\"><foreignObject width=\"61.359375\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" class=\"labelBkg\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"edgeLabel\"><p>nextItem</p></span></div></foreignObject></g></g></g><g class=\"nodes\"><g class=\"node default\" id=\"flowchart-N2-0\" transform=\"translate(533.03125, 199)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(1.1) A short introduction to Galaxy</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N3-1\" transform=\"translate(562.8046875, 363)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(1.2) Galaxy Basics for genomics</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N7-4\" transform=\"translate(141.828125, 783)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(2.3) An Introduction to Genome Assembly</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N8-5\" transform=\"translate(227.46484375, 935)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>(2.4) Chloroplast genome assembly</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N4-8\" transform=\"translate(199.0078125, 363)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-51\" width=\"260\" height=\"102\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -36)\"><rect/><foreignObject width=\"200\" height=\"72\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>Module 2: Basics of Genome Sequence Analysis</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N6-17\" transform=\"translate(231.9375, 643)\"><rect class=\"basic label-container\" style=\"\" x=\"-79.359375\" y=\"-27\" width=\"158.71875\" height=\"54\"/><g class=\"label\" style=\"\" transform=\"translate(-49.359375, -12)\"><rect/><foreignObject width=\"98.71875\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"nodeLabel\"><p>(2.2) Mapping</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N5-24\" transform=\"translate(429.890625, 515)\"><rect class=\"basic label-container\" style=\"\" x=\"-101.5859375\" y=\"-27\" width=\"203.171875\" height=\"54\"/><g class=\"label\" style=\"\" transform=\"translate(-71.5859375, -12)\"><rect/><foreignObject width=\"143.171875\" height=\"24\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table-cell; white-space: nowrap; line-height: 1.5; max-width: 200px; text-align: center;\"><span class=\"nodeLabel\"><p>(2.1) Quality Control</p></span></div></foreignObject></g></g><g class=\"node default\" id=\"flowchart-N1-32\" transform=\"translate(451.84765625, 47)\"><rect class=\"basic label-container\" style=\"\" x=\"-130\" y=\"-39\" width=\"260\" height=\"78\"/><g class=\"label\" style=\"\" transform=\"translate(-100, -24)\"><rect/><foreignObject width=\"200\" height=\"48\"><div xmlns=\"http://www.w3.org/1999/xhtml\" style=\"display: table; white-space: break-spaces; line-height: 1.5; max-width: 200px; text-align: center; width: 200px;\"><span class=\"nodeLabel\"><p>Module 1: Introduction to Galaxy</p></span></div></foreignObject></g></g></g></g></g></svg>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "#from IPython.display import display, Markdown\n",
        "#Markdown(mermaid_syntax)\n",
        "\n",
        "import base64\n",
        "from IPython.display import display_svg\n",
        "from urllib.request import Request, urlopen\n",
        "\n",
        "def mm(graph):\n",
        "    graphbytes = graph.encode(\"ascii\")\n",
        "    base64_bytes = base64.b64encode(graphbytes)\n",
        "    base64_string = base64_bytes.decode(\"ascii\")\n",
        "    url=\"https://mermaid.ink/svg/\" + base64_string\n",
        "    req=Request(url, headers={'User-Agent': 'IPython/Notebook'})\n",
        "    display_svg(urlopen(req).read().decode(), raw=True)\n",
        "mm(mermaid_syntax)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gszZ0tVRBWrW"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "base",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.12.8"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
